internals/overlord/pairingstate/manager.go

```diff
@@ -0,0 +1,312 @@
+// for deciding when to marshal a section or not.
+func (c *PairingConfig) IsZero() bool {
+	return c.Mode == ModeUnset
+}
+
+func (c *PairingConfig) Combine(other *PairingConfig) {
+	if other.Mode != ModeUnset {
+		c.Mode = other.Mode
+	}
+}
+
+type PairingManager struct {
+	state  *state.State
+	mu     sync.Mutex
+	config *PairingConfig
+	// paired is true if a previous pairing request succeeded.
+	paired bool
+	// enabled is true if the pairing window is enabled.
+	enabled bool
+	// timer controls the duration of the pairing window.
+	timer Timer
+}
+
+func NewManager(state *state.State) *PairingManager {
+	m := &PairingManager{
+		state: state,
+		config: &PairingConfig{
+			Mode: ModeUnset,
+		},
+	}
+
+	// Load the paired state at startup.
+	m.state.Lock()
+	defer m.state.Unlock()
+	m.state.Get(pairedStateKey, &m.paired)
+
+	return m
+}
+
+// PlanChanged informs the pairing manager that the plan has been updated.
+func (m *PairingManager) PlanChanged(update *plan.Plan) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	newConfig := update.Sections[PairingField].(*PairingConfig)
+
+	// If the mode changed, make sure the pairing window is disabled.
+	if m.config.Mode != newConfig.Mode {
+		m.enabled = false
+		m.stopTimer()
+	}
+	m.config = newConfig
+}
+
+// Ensure implements overlord.StateManager interface.
+func (m *PairingManager) Ensure() error {
+	return nil
+}
+
+// Stop implements overlord.StateStopper interface.
+func (m *PairingManager) Stop() {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	m.enabled = false
+	m.stopTimer()
+}
+
+// PairMTLS adds a client identity with admin permissions to the identity
+// subsystem. A pairing request always leaves the pairing window disabled.
+func (m *PairingManager) PairMTLS(clientCert *x509.Certificate) error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	if !m.enabled {
+		return errors.New("cannot pair while pairing window is disabled")
+	}
+
+	// Any success or failure should always disable the pairing window.
+	defer func() {
+		m.enabled = false
+		m.stopTimer()
+	}()
+
+	// Verify that the client certificate is self-signed (the public
+	// key included must verify the signature). We do this here as a
+	// sanity check since we are about to pair this certificate and use
+	// it persisted certificate in exactly this way for future client
+	// credential checks. Note that the TLS handshake already proved
+	// that the client has access to the private key by verifying the
+	// handshake transcript signature using the public key in this
+	// certificate.
+	roots := x509.NewCertPool()
+	roots.AddCert(clientCert)
+	opts := x509.VerifyOptions{
+		Roots: roots,
+		// We only support verifying client TLS certificates.
+		KeyUsages: []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth},
+	}
+	_, err := clientCert.Verify(opts)
+	if err != nil {
+		return fmt.Errorf("cannot verify client certificate signature: %w", err)
+	}
+
+	m.state.Lock()
+	defer m.state.Unlock()
+
+	existingIdentities := m.state.Identities()
```

Comment:
Unrelated to this PR, but it's really unfortunate that we've allowed this logic to be inserted under the state package.  Feels like every other cycle we need to go back and remove something that was added there because it felt like something central and important, when really it should be on its own subpackage.

@benhoyt Probably a topic for the sprint.